python/ray/_private/gcs_utils.py

```diff
@@ -302,6 +319,114 @@ def pin_runtime_env_uri(self, uri: str, expiration_s: int) -> None:
             )
 
 
+class GcsAioClient:
+    def __init__(
+        self,
+        channel: Optional[GcsChannel] = None,
+        address: Optional[str] = None,
+    ):
+        if channel is None:
+            assert isinstance(address, str)
+            channel = GcsChannel(gcs_address=address, aio=True)
+        assert isinstance(channel, GcsChannel)
+        assert channel._aio is True
+        self._channel = channel
+        self._connect()
+
+    def _connect(self):
+        self._channel.connect()
+        self._kv_stub = gcs_service_pb2_grpc.InternalKVGcsServiceStub(
+            self._channel.channel()
+        )
+
+    async def internal_kv_get(
+        self, key: bytes, namespace: Optional[bytes], timeout: Optional[float] = None
+    ) -> Awaitable[bytes]:
+        logger.debug(f"internal_kv_get {key} {namespace}")
+        req = gcs_service_pb2.InternalKVGetRequest(namespace=namespace, key=key)
+        reply = await self._kv_stub.InternalKVGet(req, timeout=timeout)
+        if reply.status.code == GcsCode.OK:
+            return reply.value
+        elif reply.status.code == GcsCode.NotFound:
+            return None
+        else:
+            raise RuntimeError(
```

Comment:
Is this idiomatic in Python? To throw exceptions from a function that returns an Awaitable? In folly::future, you have the thenError() method to deal with async exception propagation. The way this method is currently, the caller has be ready for a future in the normal case, None is some cases, or an exception.

Reply:
Yes, this is valid. I feel the reason right now folly::future is doing that way is because there is no coroutine support in cpp. This is more like async operations in hacklang.